OpenEBS provisioner HA

satyamz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you -->

**What this PR does / why we need it**:
    `openebs-provisioner` should be available all the time even if any node goes down.

**What is changed?**
    This will increase the `replicas` count and will add `affinity` to the pod spec.

**Which issue this PR fixes**:
fixes #267

AmitKumarDas commented on k8s/openebs-operator.yaml:
```diff
@@ -96,12 +96,20 @@ metadata:
+              matchExpressions:
+                key: node
+                operator: Equal
+                values: ["minion01","minion02"]
```
This forces the admin to have above values i.e. minion01 & minion02 in his setup.

AmitKumarDas replied:
Following is a better feature than labelling the nodes directly.
https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#inter-pod-affinity-and-anti-affinity-beta-feature

AmitKumarDas replied:
@satyamz Yes. I was pointing at podAntiAffinity with labels on Pod itself.

satyamz replied:
Yes! I am making/testing changes. Once done, will update PR

Thanks

AmitKumarDas commented on contribute/design/openebs-provisioner-high-availibility.md:
````diff
@@ -0,0 +1,77 @@
+  To do that execute: 
+
+```
+kubectl label nodes kubeminion-01  node=minion01  # For node kubeminion-01
````
It will be ideal to avoid modifications to the K8s setup. This should be the last resort.

AmitKumarDas commented on contribute/design/openebs-provisioner-high-availibility.md:
```diff
@@ -0,0 +1,77 @@
+              matchExpressions:
+                key: node			  # <-- Key of label of the node. 	
+                operator: Equal
+                values: ["minion01","minion02"]   # <-- Add the labels of the nodes as a value.
```
Will it be good to have these operator yaml generated based on the environment of the K8s setup or based on the entries provided by the Admin ?

AmitKumarDas replied:
This has nothing to do with this PR. However, see through these reviews & if required raise issues.

satyamz commented:
Added `PodAntiAffinity`  and removed `nodeAffinity` 
```yaml 

  podAntiAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          - weight: 100
            podAffinityTerm:
              labelSelector:
                matchExpressions:
                  - key: provisioner
                    operator: In
                    values:
                    - P1
                    - P2
              topologyKey: "kubernetes.io/hostname"

```

AmitKumarDas commented on k8s/openebs-operator.yaml:
```diff
@@ -99,14 +99,18 @@ spec:
-                values: ["minion01","minion02"]
+      podAntiAffinity:
+        requiredDuringSchedulingIgnoredDuringExecution:
+          - weight: 100
```
Is the weight documented somewhere ?

AmitKumarDas commented on contribute/design/openebs-provisioner-high-availibility.md:
````diff
@@ -1,44 +1,45 @@
 ```
-kubectl label nodes kubeminion-01  node=minion01  # For node kubeminion-01
-kubectl label nodes kubeminion-02  node=minion02   # For node kubeminion-02
+kubectl label pods openebs-provisioner-1149663462-6vbvm  provisioner=P2
````
1/ These steps are not required as we can set these labels in the provisioner specs directly.
2/ Let us quickly debate on the label key & label value in maya channel.
3/ Same k:v labels can be assigned into these Pods. This label(s) will repel the Pods.

kmova commented on contribute/design/openebs-provisioner-high-availibility.md:
```diff
@@ -0,0 +1,98 @@
+## Problems addressed:
+
+- `openebs-provisioner` should always be available.
+- `openebs-provisioner` should always be running on the specified/labeled pods. 
```
openebs-provisioner is itself a pod, what do you mean by running on specified/labeled pods?

kmova replied:
@satyamz -- can you clarify the above question?

satyamz replied:
@kmova It should be:  Each node should always be running the specified/labeled `openebs-provisioner`  pods on each node provided that replica count is equal to the number of nodes.

kmova replied:
@satyamz -- As per the yaml spec - kubernetes scheduler will try to run 2 pods of openebs-provisioner when it can. 

What is the behavior you expect in the following cases w.r.t to scheduler:
case1: there is only 1 node in the kubernetes cluster - with resources available to run openebs-provisioner
case2: there are 2 nodes in the kubernetes cluster - with resources available to run openebs-provisioner
case3:- there are 3 nodes in the kubernetes cluster - with resources available to run openebs-provisioner

Also, in case1, will it try to evacuate or stop --other low priority pods to make room for openebs-provisioner? 

In case the scheduler is unable to schedule openebs-provisioner, how will the user know what actions to be taken to rectify the situation.

satyamz replied:
This PR will make sure that at least one `openebs-provisioner` pod is running on each node (considering `replicas` == the number of nodes ). We are trying to solve the problem where `openebs-provisioner` pods are running on all nodes. Hence, pod anti-affinity:
> Pod anti-affinity can prevent the scheduler from locating a new pod on the same node as pods with the same labels if the label selector on the new pod matches the label on the current pod. 

Regarding above cases:
>What is the behavior you expect in the following cases w.r.t to scheduler:
case1: there is only 1 node in the kubernetes cluster - with resources available to run openebs-provisioner
case2: there are 2 nodes in the kubernetes cluster - with resources available to run openebs-provisioner
case3:- there are 3 nodes in the kubernetes cluster - with resources available to run openebs-provisioner
Also, in case1, will it try to evacuate or stop --other low priority pods to make room for openebs-provisioner?

For this, we need to have PDB & PriorityClass for `openebs-provisioner` and other OpenEBS components. I will create a new issue for these. (This was also suggested by Amit)

satyamz replied:
@kmova  created issue https://github.com/openebs/openebs/issues/737 for this and these cases will be covered in different PR.

kmova commented on contribute/design/openebs-provisioner-high-availibility.md:
```diff
@@ -0,0 +1,98 @@
+- Increase the `replicas` to 2 or >2 in `openebs-operator.yaml` for `openebs-provisioner` deployment. 
+
+
+- We need to make sure that each node is running at least one  `openebs-provisioner` pod.  
```
what is there are 3 nodes or 10 nodes, will the provisioner run on each node of the 10 nodes?

kmova commented on contribute/design/openebs-provisioner-high-availibility.md:
```diff
@@ -0,0 +1,98 @@
+
+---
+
+Once, we labeled the pods. If we change the `replicas` count to 5 then 5 pods will be running. If we descrease the replica count to 2 then only two pods with label P1 and P2 will be running and other 3 pods will be deleted. 
```
Is this section still accurate?

satyamz commented:
@kmova I haven't updated the doc yet. I will do it today.

satyamz commented:
@kmova Please check the updated document. 
Thanks

satyamz commented:
Closing this PR in favor of https://github.com/openebs/openebs/pull/814
